### PR TITLE
add java and cpp sol to cses elevator rides

### PR DESCRIPTION
**Pull request by TruaShamu**

### CODE REVIEW

File: solutions/cses-1653.mdx [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+<LanguageSection>
+
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class elevators {
```
Comment (nathangong) [inline solutions/cses-1653.mdx:20]:
```suggestion
public class ElevatorRides {
```

File: solutions/cses-1653.mdx [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+				}
+			}
+		}
+		// Result when all people have used the elevator.
+		io.println(dp[(1 << people) - 1].rides);
+		io.close();
+	}
+
+	// State (i,j) means that we're on the ith ride, and we've used up j weight.
+	static class state {
```
Comment (nathangong) [inline solutions/cses-1653.mdx:76]:
```suggestion
	static class State {
```

File: solutions/cses-1653.mdx [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+		public int rides, weight;
+
+		public state(int rides, int weight) {
```
Comment (nathangong) [inline solutions/cses-1653.mdx:79]:
```suggestion
		public State(int rides, int weight) {
```

File: solutions/cses-1653.mdx [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		}
+
+		/*
+		dp[bitmask] is the current best state
+		in terms of the number of rides used, and weight used on the last ride.
+		The bitmask is used to represent whether the people have used the elevator.
+		If the ith bit is set, it means the ith person has used the elevator.
+		*/
+		dp = new state[1 << people];
+		dp[0] = new state(1, 0);
```
Comment (nathangong) [inline solutions/cses-1653.mdx:42]:
```suggestion
		dp[0] = new State(1, 0);
```

File: solutions/cses-1653.mdx [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class elevators {
+	public static int people, maxWeight;
+	public static int[] weight;
+	public static state[] dp;
```
Comment (nathangong) [inline solutions/cses-1653.mdx:23]:
```suggestion
	public static State[] dp;
```

File: solutions/cses-1653.mdx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+			weight[i] = io.nextInt();
+		}
+
+		/*
+		dp[bitmask] is the current best state
+		in terms of the number of rides used, and weight used on the last ride.
+		The bitmask is used to represent whether the people have used the elevator.
+		If the ith bit is set, it means the ith person has used the elevator.
+		*/
+		dp = new state[1 << people];
```
Comment (nathangong) [inline solutions/cses-1653.mdx:41]:
```suggestion
		dp = new State[1 << people];
```

File: solutions/cses-1653.mdx [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+		dp[0] = new state(1, 0);
+
+		for (int mask = 1; mask < (1 << people); mask++) {
+			dp[mask] = new state(people + 1, maxWeight + 1);
```
Comment (nathangong) [inline solutions/cses-1653.mdx:45]:
```suggestion
			dp[mask] = new State(people + 1, maxWeight + 1);
```

File: solutions/cses-1653.mdx [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+			for (int i = 0; i < people; i++) {
+				// the ith person has used the elevator.
+				if ((mask & (1 << i)) > 0) {
+					/*
+					The previous state from when the
+					ith person has not used the elevator.
+					 */
+					int prev = mask ^ (1 << i);
+					state cur = new state(dp[prev].rides, dp[prev].weight);
```
Comment (nathangong) [inline solutions/cses-1653.mdx:54]:
```suggestion
					Statecur = new State(dp[prev].rides, dp[prev].weight);
```

File: solutions/cses-1653.mdx [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+	// State (i,j) means that we're on the ith ride, and we've used up j weight.
+	static class state {
+		public int rides, weight;
+
+		public state(int rides, int weight) {
+			this.rides = rides;
+			this.weight = weight;
+		}
+
+		public int compareTo(state oState) {
```
Comment (nathangong) [inline solutions/cses-1653.mdx:84]:
```suggestion
		public int compareTo(State oState) {
```

File: solutions/cses-1653.mdx [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+		for (int i = 0; i < people; i++) {
+			weight[i] = io.nextInt();
+		}
+
+		/*
+		dp[bitmask] is the current best state
+		in terms of the number of rides used, and weight used on the last ride.
+		The bitmask is used to represent whether the people have used the elevator.
+		If the ith bit is set, it means the ith person has used the elevator.
+		*/
```
Comment (nathangong) [inline solutions/cses-1653.mdx:40]:
```suggestion
		 */
```

File: solutions/cses-1653.mdx [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+		weight = new int[people];
+		for (int i = 0; i < people; i++) {
+			weight[i] = io.nextInt();
+		}
+
+		/*
+		dp[bitmask] is the current best state
+		in terms of the number of rides used, and weight used on the last ride.
+		The bitmask is used to represent whether the people have used the elevator.
+		If the ith bit is set, it means the ith person has used the elevator.
```
Comment (nathangong) [inline solutions/cses-1653.mdx:39]:
```suggestion
		 * dp[bitmask] is the current best state
		 * in terms of the number of rides used, and weight used on the last ride.
		 * The bitmask is used to represent whether the people have used the elevator.
		 * If the ith bit is set, it means the ith person has used the elevator.
```

File: solutions/cses-1653.mdx (hunk 5 above)
Comment (nathangong) [inline solutions/cses-1653.mdx:23]: Is there a need for these to be instance variables? If not, it's probably more efficient to just declare them in the main method

**Review by nathangong — CHANGES_REQUESTED** [review]

**Comment by TruaShamu** [comment]
Alright, done.
Sorry for making you fix all of my stuff for me

**Comment by nathangong** [comment]
Yep no problem

**Review by nathangong — APPROVED** [review]
java code looks good

File: solutions/cses-1653.mdx [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+	vector<int> weight(people);
+	for (int& i : weight)
+		cin >> i;
+
+	vector<pair<int, int>> dp(1 << people, { people + 1, maxWeight + 1 });
+	dp[0] = make_pair(1, 0);
+	/* Loop through all bitmasks.
+	   The bitmasks represent whether each person has used the elevator or not.
+	   If the ith bit is set, this means the ith person has used the elevator.
+	*/
```
Comment (maggie-j-liu) [inline solutions/cses-1653.mdx:112]: can you format this comment like the java one? with `*` in front of every line

**Comment by TruaShamu** [comment]
alright, done!

**Review by maggie-j-liu — APPROVED** [review]
looks good but i couldn't check the preview since the deployment is kinda broken

File: solutions/cses-1653.mdx [hunk 13]
```diff
@@ -0,0 +1,138 @@
+---
+id: cses-1653
+source: CSES
+title: Elevator Rides
+author: Sofia Yang
+---
+
+## Implementation
+
```
Comment (dongliu0426) [inline solutions/cses-1653.mdx:9]: can you link the CPH expl? section 10.5

Comment (TruaShamu) [reply solutions/cses-1653.mdx:9]: done!

**Pull request merged**